arclightning_backend/src/main.rs

```diff
@@ -0,0 +1,201 @@
+extern crate futures;
+extern crate hyper;
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate serde_json;
+extern crate toml;
+
+use futures::future;
+
+use hyper::rt::{Future, Stream};
+use hyper::service::service_fn;
+use hyper::{Body, Method, Request, Response, Server, StatusCode};
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+// after this concept is further understood, will switch to 'Either'
+type ResponseFuture = Box<Future<Item = Response<Body>, Error = hyper::Error> + Send>;
+
+// using PartialEq for unit tests
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+// Using clone in a unit test atm.  Might not be necessary
+struct Game {
+    name: String,
+    description: String,
+    genres: Vec<String>,
+    thumbnail_path: PathBuf,
+    exe_path: PathBuf,
+}
+
+fn router(request: Request<Body>) -> ResponseFuture {
+    let mut response = Response::new(Body::empty());
+
+    match (request.method(), request.uri().path()) {
+        (&Method::GET, "/api/v1/list_games") => {
+            *response.body_mut() = Body::from("test");
+        }
+        _ => {
+            *response.status_mut() = StatusCode::NOT_FOUND;
+        }
+    }
+    Box::new(future::ok(response))
+}
+
+fn toml_to_hashmap(toml_filepath: PathBuf) -> HashMap<String, Game> {
+    let mut games_toml = String::new();
+    let mut file = File::open(&toml_filepath).expect("Could not find .toml file");
+    file.read_to_string(&mut games_toml)
+        .unwrap_or_else(|err| panic!("Error while reading .toml file: [{}]", err));
+    toml::from_str(&games_toml).unwrap()
```

Comment:
This one actually complicates things. The above 2 failable functions return `Result<_, std::io::Error>`, while this returns `Result<_, toml::de::Error>`. Use `map_err` and `io::Error::new` with `io::ErrorKind::Other` to convert `toml::de::Error` to an `io::Error`.




sidenote:
The technically "Correct" way to do this is to create a new error type and implement `From<io::Error>` and `From<toml::de::Error>` or use something like the `failure` crate, but casting to `io::Error` is easy and works well enough for this fairly simple case.